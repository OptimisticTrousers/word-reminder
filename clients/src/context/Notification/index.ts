export { NOTIFICATION_ACTIONS, NotificationContext } from "./Context";
export { NotificationProvider } from "./Provider";
