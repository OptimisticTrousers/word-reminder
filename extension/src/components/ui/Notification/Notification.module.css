.notification {
  border-radius: 6px;
  padding: 1rem;
}

.notification--success {
  --bkg-color: rgb(220, 252, 231);
  --base-text-color: rgb(20, 83, 45);
  --btn-text-color: rgb(22, 163, 74);
  --icon-color: rgb(21, 128, 61);
  background-color: var(--bkg-color);
}

.notification--error {
}

.notification__flex {
  display: flex;
  align-items: center;
}

.notification__item {
  flex-shrink: 0;
}

.notification__box {
  margin-left: 0.75rem;
}

.notification__message {
  color: var(--base-text-color);
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.25rem;
  display: block;
}

.notification__controls {
  margin-left: auto;
  padding-left: 0.75rem;
}

.notification__control {
  margin-top: -0.375rem;
  margin-right: -0.375rem;
  margin-bottom: -0.375rem;
  margin-left: -0.375rem;
}

.notification__button {
  color: var(--btn-text-color);
  padding: 0.375rem;
  background-color: var(--bkg-color);
  border-radius: 6px;
  display: inline-flex;
}

.notification__text {
  position: absolute;
  width: 1px;
  padding: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
}

.notification__icon {
  width: 1.25rem;
  height: 1.25rem;
}

.notification__icon--circle {
  color: var(--icon-color);
}
