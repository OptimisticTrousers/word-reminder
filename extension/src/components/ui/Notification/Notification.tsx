import { CircleCheck, CircleX, X } from "lucide-react";
import { useContext } from "react";
import CSSModules from "react-css-modules";

import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import styles from "./Notification.module.css";

export const Notification = CSSModules(
  function ({ type, message }: { type: string; message: string }) {
    const { dismissNotification } = useContext(NotificationContext);

    function handleClick() {
      dismissNotification();
    }

    return (
      <div
        styleName={`notification notification--${type}`}
        role="dialog"
        aria-labelledby="notification-message"
      >
        <div styleName="notification__flex">
          <div styleName="notification__item">
            {type === NOTIFICATION_ACTIONS.SUCCESS ? (
              <CircleCheck
                data-testid="check-icon"
                styleName="notification__icon notification__icon--circle"
              />
            ) : (
              <CircleX
                data-testid="x-icon"
                styleName="notification__icon notification__icon--circle"
              />
            )}
          </div>
          <div styleName="notification__box">
            <span id="notification-message" styleName="notification__message">
              {message}
            </span>
          </div>
          <div styleName="notification__controls">
            <div styleName="notification__control">
              <button styleName="notification_button" onClick={handleClick}>
                <span styleName="notification__text">Dismiss</span>
                <X styleName="notification__icon" />
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
