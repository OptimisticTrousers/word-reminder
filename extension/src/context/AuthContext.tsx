import { createContext, FC, useReducer } from "react";

interface IUser {
  _id: string;
  username: string;
  theme: string;
  words: IUserWord[];
}

interface IWord {
  _id: string;
  word: string;
  origin: string;
  phonetic: string;
  audio: string;
  meanings: [
    {
      partOfSpeech: string;
      definitions: [
        {
          definition: string;
          example: string;
          synonyms: string[];
          antonyms: string[];
        }
      ];
    }
  ];
}

export interface IUserWord {
  _id: string;
  userId: IUser | string;
  word: IWord;
  learned: boolean;
}

interface AuthState {
  user: IUser | null;
}

interface Action {
  type: string;
  payload: IUser;
}

interface AuthContext {
  state: AuthState;
  dispatch: (action: Action) => void;
}

const authReducer = (state: AuthState, action: Action) => {
  switch (action.type) {
    // Login action will always contain a payload of user object
    case "LOGIN":
      return { user: action.payload };
    // Remove user object on logout action dispatch
    case "LOGOUT":
      return { user: null };
    default:
      return state;
  }
};

export const AuthContext = createContext<AuthContext>({} as AuthContext);

interface Props {
  children: JSX.Element[] | JSX.Element;
}

const AuthProvider: FC<Props> = ({ children }) => {
  const [state, dispatch] = useReducer(authReducer, {
    user: null,
  });

  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;
