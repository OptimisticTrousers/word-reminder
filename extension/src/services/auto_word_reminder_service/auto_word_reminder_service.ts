import { AutoWordReminderParams } from "common";

import { service } from "../service";

export const autoWordReminderService = (function (service) {
  const { get, post, put, remove, VITE_API_DOMAIN } = service;

  function getAutoWordReminder(userId: string) {
    return get({
      url: `${VITE_API_DOMAIN}/users/${userId}/autoWordReminders`,
      options: {
        credentials: "include",
      },
    });
  }

  function createAutoWordReminder({
    userId,
    body,
  }: {
    userId: string;
    body: Omit<AutoWordReminderParams, "user_id">;
  }) {
    return post({
      url: `${VITE_API_DOMAIN}/users/${userId}/autoWordReminders`,
      options: { body: JSON.stringify(body), credentials: "include" },
    });
  }

  function updateAutoWordReminder({
    userId,
    autoWordReminderId,
    body,
  }: {
    userId: string;
    autoWordReminderId: string;
    body: Omit<AutoWordReminderParams, "user_id">;
  }) {
    return put({
      url: `${VITE_API_DOMAIN}/users/${userId}/autoWordReminders/${autoWordReminderId}`,
      options: { body: JSON.stringify(body), credentials: "include" },
    });
  }

  function deleteAutoWordReminder({
    userId,
    autoWordReminderId,
  }: {
    userId: string;
    autoWordReminderId: string;
  }) {
    return remove({
      url: `${VITE_API_DOMAIN}/users/${userId}/autoWordReminders/${autoWordReminderId}`,
      options: { credentials: "include" },
    });
  }

  return {
    createAutoWordReminder,
    getAutoWordReminder,
    deleteAutoWordReminder,
    updateAutoWordReminder,
  };
})(service);
