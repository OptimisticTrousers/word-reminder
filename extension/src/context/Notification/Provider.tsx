import { useReducer, ReactNode } from "react";

import { NOTIFICATION_ACTIONS, NotificationContext } from "./Context";
import { Notification } from "../../components/ui/Notification/Notification";

interface Props {
  children: ReactNode;
}

interface NotificationState {
  isVisible: boolean;
  type: string;
  message: string;
}

interface NotificationAction {
  payload: { message: string };
  type: string;
}

export interface NotificationContext {
  showNotification: (type: string, message: string) => void;
  dismissNotification: () => void;
}

const notificationReducer = (
  state: NotificationState,
  action: NotificationAction
) => {
  switch (action.type) {
    case NOTIFICATION_ACTIONS.SUCCESS:
      return {
        ...state,
        isVisible: true,
        type: NOTIFICATION_ACTIONS.SUCCESS,
        message: action.payload.message,
      };
    case NOTIFICATION_ACTIONS.ERROR:
      return {
        ...state,
        isVisible: true,
        type: NOTIFICATION_ACTIONS.ERROR,
        message: action.payload.message,
      };
    case DISAPPEAR:
      return { ...state, isVisible: false, type: "", message: "" };
    default: {
      throw Error("Unknown action: " + action.type);
    }
  }
};

// Provide ability to spawn Toast notification globally within the application
export function NotificationProvider({ children }: Props) {
  const [state, dispatch] = useReducer(notificationReducer, {
    isVisible: false,
    type: "",
    message: "",
  });

  function dismissNotification() {
    dispatch({
      type: DISAPPEAR,
      payload: { message: "" },
    });
  }

  function showNotification(type: string, message: string): void {
    const delay = 5000;

    dispatch({ type, payload: { message } });

    setTimeout(() => {
      dismissNotification();
    }, delay);
  }

  const { isVisible, type, message } = state;

  return (
    <NotificationContext.Provider
      value={{ showNotification, dismissNotification }}
    >
      {isVisible && <Notification type={type} message={message} />}
      {children}
    </NotificationContext.Provider>
  );
}

const DISAPPEAR = "disappear";
