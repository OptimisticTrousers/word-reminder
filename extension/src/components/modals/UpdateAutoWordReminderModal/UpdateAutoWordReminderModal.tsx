import { AutoWordReminder, SortMode, User } from "common";
import CSSModules from "react-css-modules";
import { useOutletContext } from "react-router-dom";
import { useNotificationError } from "../../../hooks/useNotificationError";
import { ToggleModal } from "../types";
import { useMutation, useQueryClient } from "@tanstack/react-query";
import { useContext } from "react";
import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import { ModalContainer } from "../ModalContainer";
import styles from "./UpdateAutoWordReminderModal.module.css";
import { Reminder } from "../../ui/Reminder";
import { Duration } from "../../ui/Duration";
import { autoWordReminderService } from "../../../services/auto_word_reminder_service/auto_word_reminder_service";
import { msToUnits, unitsToMs } from "../../../utils/date/date";

interface Props {
  toggleModal: ToggleModal;
  autoWordReminder: AutoWordReminder;
}

export const UpdateAutoWordReminderModal = CSSModules(
  function ({ toggleModal, autoWordReminder }: Props) {
    const { user }: { user: User } = useOutletContext();
    const userId = String(user.id);
    const { showNotification } = useContext(NotificationContext);
    const { showNotificationError } = useNotificationError();
    const queryClient = useQueryClient();
    const { isPending, mutate } = useMutation({
      mutationFn: autoWordReminderService.updateAutoWordReminder,
      onSuccess: () => {
        showNotification(
          NOTIFICATION_ACTIONS.SUCCESS,
          UPDATE_AUTO_WORD_REMINDER_NOTIFICATION_MSGS.updateAutoWordReminder()
        );
        queryClient.invalidateQueries({
          queryKey: ["wordReminders"],
        });
        queryClient.invalidateQueries({
          queryKey: ["autoWordReminders"],
          exact: true,
        });
      },
      onError: showNotificationError,
      onSettled: () => {
        toggleModal();
      },
    });

    function handleUpdate(formData: FormData) {
      const time = {
        minutes: Number(formData.get("minutes") as string),
        hours: Number(formData.get("hours") as string),
        days: Number(formData.get("days") as string),
        weeks: Number(formData.get("weeks") as string),
      };

      mutate({
        userId,
        autoWordReminderId: String(autoWordReminder.id),
        body: {
          reminder: formData.get("reminder") as string,
          create_now: (formData.get("create_now") as string) === "on",
          duration: unitsToMs(time),
          word_count: Number(formData.get("word_count") as string),
          is_active: (formData.get("is_active") as string) === "on",
          has_reminder_onload:
            (formData.get("has_reminder_onload") as string) === "on",
          has_learned_words:
            (formData.get("has_learned_words") as string) === "on",
          sort_mode: formData.get("sort_mode") as SortMode,
        },
      });
    }

    return (
      <ModalContainer
        title="Update Auto Word Reminder"
        toggleModal={toggleModal}
      >
        <form styleName="modal__form" action={handleUpdate}>
          <Reminder disabled={false} value={autoWordReminder.reminder} />
          <Duration
            disabled={false}
            defaultValues={msToUnits(autoWordReminder.duration)}
          />
          <div styleName="modal__control">
            <label styleName="modal__label" htmlFor="word_count">
              Word Count
            </label>
            <input
              styleName="modal__input"
              type="number"
              id="word_count"
              name="word_count"
              required
              max={99}
              defaultValue={autoWordReminder.word_count}
            />
          </div>
          <fieldset styleName="modal__fieldset">
            <legend styleName="modal__legend">Options</legend>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="create_now">
                Create Now
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="create_now"
                name="create_now"
                defaultChecked
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="is_active">
                Is Active
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="is_active"
                name="is_active"
                defaultChecked={autoWordReminder.is_active}
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="has_reminder_onload">
                Has Reminder Onload
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="has_reminder_onload"
                name="has_reminder_onload"
                defaultChecked={autoWordReminder.has_reminder_onload}
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="has_learned_words">
                Has Learned Words
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="has_learned_words"
                name="has_learned_words"
                defaultChecked={autoWordReminder.has_learned_words}
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="sort_mode">
                Sort Mode
              </label>
              <select
                styleName="modal__select"
                name="sort_mode"
                id="sort_mode"
                required
                defaultValue={autoWordReminder.sort_mode}
              >
                <option styleName="modal__option" value={SortMode.Newest}>
                  Newest
                </option>
                <option styleName="modal__option" value={SortMode.Oldest}>
                  Oldest
                </option>
                <option styleName="modal__option" value={SortMode.Random}>
                  Random
                </option>
              </select>
            </div>
          </fieldset>
          <button styleName="modal__button" type="submit" disabled={isPending}>
            Update
          </button>
        </form>
      </ModalContainer>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);

const UPDATE_AUTO_WORD_REMINDER_NOTIFICATION_MSGS = {
  updateAutoWordReminder: () => {
    return "Your auto word reminder preferences have been updated!";
  },
};
