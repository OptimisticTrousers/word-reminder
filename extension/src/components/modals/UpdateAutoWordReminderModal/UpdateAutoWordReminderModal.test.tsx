import { render, screen } from "@testing-library/react";
import { createRoutesStub, Outlet } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import { UpdateAutoWordReminderModal } from "./UpdateAutoWordReminderModal";
import userEvent from "@testing-library/user-event";
import { SortMode } from "common";
import { NotificationProvider } from "../../../context/Notification";
import { Mock } from "vitest";
import { autoWordReminderService } from "../../../services/auto_word_reminder_service/auto_word_reminder_service";

vi.mock("../ModalContainer/ModalContainer");

describe("UpdateAutoWordReminderModal component", () => {
  const testUser = {
    id: 1,
  };

  const autoWordReminder = {
    user_id: testUser.id,
    id: 1,
    reminder: "* * * * *",
    create_now: false,
    duration: 36000000,
    word_count: 7,
    is_active: true,
    has_reminder_onload: true,
    has_learned_words: false,
    sort_mode: SortMode.Oldest,
    created_at: new Date(),
    updated_at: new Date(),
  };

  function setup({
    toggleModal,
    queryClient,
  }: {
    toggleModal: Mock;
    queryClient: QueryClient;
  }) {
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <NotificationProvider>
                  <QueryClientProvider client={queryClient}>
                    <UpdateAutoWordReminderModal
                      autoWordReminder={autoWordReminder}
                      toggleModal={toggleModal}
                    />
                  </QueryClientProvider>
                </NotificationProvider>
              );
            },
          },
        ],
      },
    ]);

    return {
      user: userEvent.setup(),
      ...render(<Stub initialEntries={["/"]} />),
    };
  }

  const status = 200;

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders form with all inputs", async () => {
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const { asFragment } = setup({ toggleModal: mockToggleModal, queryClient });

    const reminder = screen.getByText("Reminder");
    const duration = screen.getByText("Duration");
    const wordCount = screen.getByLabelText("Word Count", {
      selector: "input",
    });
    const createNow = screen.getByLabelText("Create Now", {
      selector: "input",
    });
    const isActive = screen.getByLabelText("Is Active", { selector: "input" });
    const hasReminderOnload = screen.getByLabelText("Has Reminder Onload", {
      selector: "input",
    });
    const hasLearnedWords = screen.getByLabelText("Has Learned Words", {
      selector: "input",
    });
    const sortMode = screen.getByLabelText("Sort Mode", { selector: "select" });
    const notification = screen.queryByRole("dialog");
    const modalHeading = screen.getByTestId("modal-heading");
    expect(modalHeading).toBeInTheDocument();
    expect(modalHeading).toHaveTextContent("Update Auto Word Reminder");
    expect(notification).not.toBeInTheDocument();
    expect(reminder).toBeInTheDocument();
    expect(duration).toBeInTheDocument();
    expect(wordCount).toBeInTheDocument();
    expect(createNow).toBeInTheDocument();
    expect(createNow).toBeChecked();
    expect(isActive).toBeInTheDocument();
    expect(isActive).toBeChecked();
    expect(hasReminderOnload).toBeInTheDocument();
    expect(hasReminderOnload).toBeChecked();
    expect(hasLearnedWords).toBeInTheDocument();
    expect(hasLearnedWords).not.toBeChecked();
    expect(sortMode).toBeInTheDocument();
    expect(sortMode).toHaveValue(SortMode.Oldest);
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls the functions to update an auto word reminder", async () => {
    const mockUpdateAutoWordReminder = vi
      .spyOn(autoWordReminderService, "updateAutoWordReminder")
      .mockImplementation(async () => {
        return {
          json: { autoWordReminder },
          status,
        };
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });
    const reminder = screen.getByLabelText("Reminder");
    const minutes = screen.getByLabelText("Minutes");
    const hours = screen.getByLabelText("Hours");
    const days = screen.getByLabelText("Days");
    const weeks = screen.getByLabelText("Weeks");
    const wordCount = screen.getByLabelText("Word Count", {
      selector: "input",
    });
    const updateButton = screen.getByRole("button", { name: "Update" });
    const createNow = screen.getByLabelText("Create Now", {
      selector: "input",
    });
    const isActive = screen.getByLabelText("Is Active", { selector: "input" });
    const hasReminderOnload = screen.getByLabelText("Has Reminder Onload", {
      selector: "input",
    });
    const hasLearnedWords = screen.getByLabelText("Has Learned Words", {
      selector: "input",
    });
    const sortMode = screen.getByLabelText("Sort Mode", { selector: "select" });
    await user.clear(reminder);
    await user.clear(minutes);
    await user.clear(hours);
    await user.clear(days);
    await user.clear(weeks);
    await user.clear(wordCount);

    await user.type(reminder, "* * * * *");
    await user.type(minutes, "1");
    await user.type(hours, "7");
    await user.type(days, "1");
    await user.type(weeks, "1");
    await user.type(wordCount, "7");
    await user.click(createNow);
    await user.click(isActive);
    await user.click(hasReminderOnload);
    await user.click(hasLearnedWords);
    await user.selectOptions(sortMode, SortMode.Oldest);
    await user.click(updateButton);

    const notification = screen.getByRole("dialog", {
      name: "Your auto word reminder preferences have been updated!",
    });
    expect(notification).toBeInTheDocument();
    expect(mockInvalidateQueries).toHaveBeenCalledTimes(1);
    expect(mockInvalidateQueries).toHaveBeenCalledWith({
      queryKey: ["autoWordReminders"],
      exact: true,
    });
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
    expect(mockUpdateAutoWordReminder).toHaveBeenCalledTimes(1);
    expect(mockUpdateAutoWordReminder).toHaveBeenCalledWith({
      userId: String(testUser.id),
      autoWordReminderId: String(autoWordReminder.id),
      body: {
        reminder: "* * * * *",
        create_now: false,
        duration: 716460000, // 1 week, 1 day, 7 hours, 1 minute in milliseconds
        word_count: 7,
        is_active: !autoWordReminder.is_active,
        has_reminder_onload: !autoWordReminder.has_reminder_onload,
        has_learned_words: !autoWordReminder.has_learned_words,
        sort_mode: SortMode.Oldest,
      },
    });
  });

  it("calls the functions to show a notification error", async () => {
    const message = "Bad Request.";
    const status = 400;
    const mockUpdateAutoWordReminder = vi
      .spyOn(autoWordReminderService, "updateAutoWordReminder")
      .mockImplementation(async () => {
        return Promise.reject({ json: { message }, status });
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });
    const updateButton = screen.getByRole("button", { name: "Update" });

    await user.click(updateButton);

    const notification = screen.getByRole("dialog", { name: message });
    expect(notification).toBeInTheDocument();
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
    expect(mockUpdateAutoWordReminder).toHaveBeenCalledTimes(1);
    expect(mockUpdateAutoWordReminder).toHaveBeenCalledWith({
      userId: String(testUser.id),
      autoWordReminderId: String(autoWordReminder.id),
      body: {
        reminder: "* * * * *",
        duration: 36000000, // 6 hours in milliseconds
        word_count: autoWordReminder.word_count,
        is_active: autoWordReminder.is_active,
        create_now: true,
        has_reminder_onload: autoWordReminder.has_reminder_onload,
        has_learned_words: autoWordReminder.has_learned_words,
        sort_mode: autoWordReminder.sort_mode,
      },
    });
  });

  it("disables the create button when the mutation is loading", async () => {
    const delay = 50;
    const mockUpdateAutoWordReminder = vi
      .spyOn(autoWordReminderService, "updateAutoWordReminder")
      .mockImplementation(async () => {
        return new Promise((resolve) => {
          setTimeout(() => {
            resolve({ json: { autoWordReminder }, status });
          }, delay);
        });
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });
    const updateButton = screen.getByRole("button", { name: "Update" });

    await user.click(updateButton);

    const notification = screen.queryByRole("dialog");
    expect(notification).not.toBeInTheDocument();
    expect(updateButton).toBeDisabled();
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
    expect(mockToggleModal).not.toHaveBeenCalled();
    expect(mockUpdateAutoWordReminder).toHaveBeenCalledTimes(1);
    expect(mockUpdateAutoWordReminder).toHaveBeenCalledWith({
      userId: String(testUser.id),
      autoWordReminderId: String(autoWordReminder.id),
      body: {
        reminder: "* * * * *",
        duration: 36000000, // 6 hours in milliseconds
        word_count: autoWordReminder.word_count,
        is_active: autoWordReminder.is_active,
        create_now: true,
        has_reminder_onload: autoWordReminder.has_reminder_onload,
        has_learned_words: autoWordReminder.has_learned_words,
        sort_mode: autoWordReminder.sort_mode,
      },
    });
  });

  it("calls the functions to close a modal", async () => {
    const mockUpdateAutoWordReminder = vi
      .spyOn(autoWordReminderService, "updateAutoWordReminder")
      .mockImplementation(async () => {
        return { json: { autoWordReminder }, status };
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });

    const closeButton = screen.getByRole("button", { name: "Close modal" });
    await user.click(closeButton);

    const notification = screen.queryByRole("dialog");
    expect(notification).not.toBeInTheDocument();
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
    expect(mockUpdateAutoWordReminder).not.toHaveBeenCalled();
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
  });

  describe("form validation", () => {
    it("only allows the user to enter up to 99 for wordCount", async () => {
      const mockUpdateAutoWordReminder = vi
        .spyOn(autoWordReminderService, "updateAutoWordReminder")
        .mockImplementation(async () => {
          return { json: { autoWordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient();
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const wordCount = screen.getByLabelText("Word Count", {
        selector: "input",
      });
      const updateButton = screen.getByRole("button", { name: "Update" });

      await user.clear(wordCount);
      await user.type(wordCount, "100");
      await user.click(updateButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(mockUpdateAutoWordReminder).not.toHaveBeenCalled();
      expect(mockToggleModal).not.toHaveBeenCalled();
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });

    it("does not allow the user to submit when wordCount is empty", async () => {
      const mockUpdateAutoWordReminder = vi
        .spyOn(autoWordReminderService, "updateAutoWordReminder")
        .mockImplementation(async () => {
          return { json: { autoWordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient();
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const wordCount = screen.getByLabelText("Word Count", {
        selector: "input",
      });
      const updateButton = screen.getByRole("button", { name: "Update" });

      await user.clear(wordCount);
      await user.clear(wordCount);
      await user.click(updateButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(mockUpdateAutoWordReminder).not.toHaveBeenCalled();
      expect(mockToggleModal).not.toHaveBeenCalled();
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });
  });
});
