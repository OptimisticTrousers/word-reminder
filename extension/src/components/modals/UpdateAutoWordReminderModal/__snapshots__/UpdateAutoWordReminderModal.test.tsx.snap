// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UpdateAutoWordReminderModal component > renders form with all inputs 1`] = `
<DocumentFragment>
  <div>
    <h2
      data-testid="modal-heading"
    >
      Update Auto Word Reminder
    </h2>
    <button
      aria-label="Close modal"
    >
      X
    </button>
    <div>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="_modal__form_c266e3"
      >
        <div
          class="_reminder_a2fc7e"
        >
          <a
            href="https://cronprompt.com/"
          >
            Visit cronprompt for more details
          </a>
          <p
            class="_reminder__description_a2fc7e"
          >
            Every minute
          </p>
          <label
            class="_reminder__label_a2fc7e"
          >
            Reminder
            <input
              class="_reminder__input_a2fc7e"
              name="reminder"
              placeholder="Enter a valid cron expression (ie. * * * * * or */5 * * * *)"
              required=""
              type="text"
              value="* * * * *"
            />
          </label>
        </div>
        <fieldset
          class="_duration_03697c"
        >
          <legend
            class="_duration__legend_03697c"
          >
            Duration
          </legend>
          <div
            class="_duration__control_03697c"
          >
            <label
              class="_duration__label_03697c"
              for="minutes"
            >
              Minutes
            </label>
            <input
              class="_duration__input_03697c"
              id="minutes"
              max="59"
              min="0"
              name="minutes"
              required=""
              type="number"
              value="0"
            />
          </div>
          <div
            class="_duration__control_03697c"
          >
            <label
              class="_duration__label_03697c"
              for="hours"
            >
              Hours
            </label>
            <input
              class="_duration__input_03697c"
              id="hours"
              max="23"
              min="0"
              name="hours"
              required=""
              type="number"
              value="10"
            />
          </div>
          <div
            class="_duration__control_03697c"
          >
            <label
              class="_duration__label_03697c"
              for="days"
            >
              Days
            </label>
            <input
              class="_duration__input_03697c"
              id="days"
              max="30"
              min="0"
              name="days"
              required=""
              type="number"
              value="0"
            />
          </div>
          <div
            class="_duration__control_03697c"
          >
            <label
              class="_duration__label_03697c"
              for="weeks"
            >
              Weeks
            </label>
            <input
              class="_duration__input_03697c"
              id="weeks"
              max="51"
              min="0"
              name="weeks"
              required=""
              type="number"
              value="0"
            />
          </div>
        </fieldset>
        <div
          class="_modal__control_c266e3"
        >
          <label
            class="_modal__label_c266e3"
            for="word_count"
          >
            Word Count
          </label>
          <input
            class="_modal__input_c266e3"
            id="word_count"
            max="99"
            name="word_count"
            required=""
            type="number"
            value="7"
          />
        </div>
        <fieldset
          class="_modal__fieldset_c266e3"
        >
          <legend
            class="_modal__legend_c266e3"
          >
            Options
          </legend>
          <div
            class="_modal__control_c266e3"
          >
            <label
              class="_modal__label_c266e3"
              for="create_now"
            >
              Create Now
            </label>
            <input
              checked=""
              class="_modal__input_c266e3"
              id="create_now"
              name="create_now"
              type="checkbox"
            />
          </div>
          <div
            class="_modal__control_c266e3"
          >
            <label
              class="_modal__label_c266e3"
              for="is_active"
            >
              Is Active
            </label>
            <input
              checked=""
              class="_modal__input_c266e3"
              id="is_active"
              name="is_active"
              type="checkbox"
            />
          </div>
          <div
            class="_modal__control_c266e3"
          >
            <label
              class="_modal__label_c266e3"
              for="has_reminder_onload"
            >
              Has Reminder Onload
            </label>
            <input
              checked=""
              class="_modal__input_c266e3"
              id="has_reminder_onload"
              name="has_reminder_onload"
              type="checkbox"
            />
          </div>
          <div
            class="_modal__control_c266e3"
          >
            <label
              class="_modal__label_c266e3"
              for="has_learned_words"
            >
              Has Learned Words
            </label>
            <input
              class="_modal__input_c266e3"
              id="has_learned_words"
              name="has_learned_words"
              type="checkbox"
            />
          </div>
          <div
            class="_modal__control_c266e3"
          >
            <label
              class="_modal__label_c266e3"
              for="sort_mode"
            >
              Sort Mode
            </label>
            <select
              class="_modal__select_c266e3"
              id="sort_mode"
              name="sort_mode"
              required=""
            >
              <option
                class="_modal__option_c266e3"
                value="newest"
              >
                Newest
              </option>
              <option
                class="_modal__option_c266e3"
                selected=""
                value="oldest"
              >
                Oldest
              </option>
              <option
                class="_modal__option_c266e3"
                value="random"
              >
                Random
              </option>
            </select>
          </div>
        </fieldset>
        <button
          class="_modal__button_c266e3"
          type="submit"
        >
          Update
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
